src/esm_parser/provenance.py

```diff
@@ -0,0 +1,842 @@
+        provenance history with functions that modified it.
+
+        Parameters
+        ----------
+        provenance_step : dict
+            Provenance entry of the current step
+        func : str
+            Function triggering this method
+        modified_by : str
+            Name of the key for the labelling the type of modification
+
+        Returns
+        -------
+        provenance_step : dict
+            Provenance entry of the current step with the ``modified_by`` item
+        """
+        if provenance_step is not None:
+            provenance_step[modified_by] = str(func)
+
+        return provenance_step
+
+
+# ========================================================
+# PROVENANCE WRAPPER FACTORY CLASS METHODS AND PROPERTIES
+# ========================================================
+@classmethod
+def wrapper_with_provenance_new(cls, *args, **kwargs):
+    """
+    To be used as the ``__new__`` method for WithProvenance classes. This is key for
+    ``copy.deepcopy``, without this ``copy.deepcopy`` breaks.
+    """
+    return super(cls, cls).__new__(cls, args[1])
+
+
+def wrapper_with_provenance_init(self, value, provenance=None):
+    """
+    To be used as the ``__init__`` method for WithProvenance classes. Adds the
+    ``provenance`` value as an instance of ``Provenance`` to the ``self._provenance``
+    attribute, and stores the original ``value`` to the ``self.value`` attribute.
+
+    Parameters
+    ----------
+    value : any
+        Value of the object
+    provenance : any
+        The provenance information
+    """
+    self._provenance = Provenance(provenance)
+    self.value = value
+
+
+@property
+def prop_provenance(self):
+    """
+    To be used as the ``provenance`` property in WithProvenance classes.
+
+    Returns
+    -------
+    self._provenance : esm_parser.provenance.Provenance
+        The provenance history stored in ``self._provenance``
+    """
+    return self._provenance
+
+
+@prop_provenance.setter
+def prop_provenance(self, new_provenance):
+    """
+    Setter for the ``provenance`` property of WithProvenance classes. Makes sure that
+    any value assigned to this property is a ``Provenance`` object and if it is not
+    returns an error.
+
+    Parameters
+    ----------
+    new_provenance : esm_parser.provenance.Provenance
+        New provenance history to be set
+
+    Raises
+    ------
+    ValueError :
+        If the given ``new_provenance`` is not a ``Provenance`` object
+    """
+    # Check if new_provenance is an instance of Provenance
+    if not isinstance(new_provenance, Provenance):
+        raise ValueError(
+            "Provenance must be an instance of the provenance.Provenance class!"
+        )
+
+    self._provenance = new_provenance
+
+
+# =======================================================
+# CLASSES FOR THE UNSUBCLASSABLE CLASSES (BOOL AND NONE)
+# =======================================================
+class ProvenanceClassForTheUnsubclassable:
```

Comment:
This makes me cry on the inside.

Reply:
Me too... I feel like I have mad a pact with the devil... but... it works so let's hope we don't burn in hell in a few months